Add geodetic representations

astrojuanlu opened the pull request:
### Description

In line with discussions at #4261, this pull request aims to add geodetic representations.

I started with `WGS84GeodeticRepresentation` to touch base and get a few eyes on my approach. In particular, I don't think we can cleanly have a generic `GeodeticRepresentation` where the ellipsoid is an extra parameter, retained in roundtrip transformations, but I would be glad to be proven wrong.

The tests are almost a copy-paste of `TestSphericalRepresentation`, with a few changes:

- `height` instead of `distance`
- `height` can only be a `Quantity`, not a `Distance`
- I didn't add differentials (not sure if worth it)

Besides, I also adapted `test_gc2gd` and `test_gd2gc` from `test_earth.py` to make sure we get the same results.

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -2222,6 +2222,62 @@ def to_cartesian(self):
+        # TODO: Check ret?
+        lon, lat, height, ret = erfa_ufunc.gc2gd(1, cart.get_xyz(xyz_axis=-1).to_value(u.m))
+        return cls(
+            Longitude(lon * u.radian, u.degree,
```
While you are at it: do change this to `lon << u.radian`, etc., that avoids an extra copy.

astrojuanlu replied:
Good catch!

mhvk reviewed (commented):
@astrojuanlu - this looks good! I'm actually pleasantly surprised to see how relatively simple it is to add - the work is mostly in the tests!

Beyond the small in-line comments: I think these representations belong in `earth.py` as they are not useful as general representations. Similarly, I'd put the tests in `test_earth`. (Or perhaps a new `earth_representations` or even an `earth/` module. Though seems a bit overkill right now.)

I agree that for now we can ignore velocities - we can add the ability to mimic observations from airplanes when the need arises!

How were you thinking of proceeding? A first step might simply be to remove the transformation machinery in `EarthLocation` and instead go through these representations inside `from_geodetic` and `lon, lat, heigh` - `to_geodetic()` may need some kind of deprecation warning as right now it returns a `namedtuple` and people may count on it. (I think that is OK, we should move to `.represent_as` anyway, making it more frame-like.)

p.s. I'll change this to draft right now just to be sure we don't accidentally merge.

astrojuanlu commented on astropy/coordinates/representation.py:
```diff
@@ -2222,6 +2222,62 @@ def to_cartesian(self):
+        cartesian coordiantes.
+        """
+        # TODO: Check ret?
+        xyz, ret = erfa_ufunc.gd2gc(1,
```
Good point about the class attribute, I'll explore that!

About `erfa.gd2gc1` though: I can't seem to find it :question:

mhvk replied:
That was a type (1 instead of back tick), i.e., just use `erfa.gd2gc`. Sorry!

astrojuanlu replied:
Oh, got it!

astrojuanlu commented on astropy/coordinates/representation.py:
```diff
@@ -2222,6 +2222,62 @@ def to_cartesian(self):
+
+    attr_classes = {'lon': Longitude,
+                    'lat': Latitude,
+                    'height': u.Quantity}
```
Will do! I was dissuaded from doing it because I didn't want the parallax, cosmology, etc But it's good to enforce the unit here.

mhvk replied:
Indeed, maybe we should have a `Length` class or so... But my suggestion was indeed only to enforce the unit, which seems easiest through the class.

astrojuanlu replied:
Besides, notice that in other representation classes, `Distance` is not used in `attr_classes`, but in the initializer. See for example:

https://github.com/astropy/astropy/blob/0ff46cb5305d288af47bd3de4432c7dc941b33cb/astropy/coordinates/representation.py#L1829-L1848

astrojuanlu replied:
...which, interestingly, doesn't work: https://github.com/astropy/astropy/issues/11104

astrojuanlu commented:
Thanks for the prompt review @mhvk !

I agree with moving this to `earth.py` and with all your comments. I wanted to think about the way forward with `EarthLocation` today but I couldn't, so I don't have any ideas yet - will give it some thought tomorrow.

mkbrewer commented:
> I agree that for now we can ignore velocities - we can add the ability to mimic observations from airplanes when the need arises!

Don't forget balloons!  https://askdrphilastro.wordpress.com/2019/12/15/blast-antarctica-2019-2020/

mhvk commented:
> Don't forget balloons!

How can I forget my own colleague!  So, we will have to get there...

astrojuanlu commented on astropy/coordinates/tests/test_earth.py:
```diff
@@ -7,16 +7,43 @@
 from astropy.coordinates.name_resolve import NameResolveError
 
 
+def components_equal(rep1, rep2):
+    result = True
+    if type(rep1) is not type(rep2):
+        return False
+    for component in rep1.components:
+        result &= getattr(rep1, component) == getattr(rep2, component)
+    return result
+
+
+def representation_equal(rep1, rep2):
+    result = True
+    if type(rep1) is not type(rep2):
+        return False
+    if getattr(rep1, '_differentials', False):
+        if rep1._differentials.keys() != rep2._differentials.keys():
+            return False
+        for key, diff1 in rep1._differentials.items():
+            result &= components_equal(diff1, rep2._differentials[key])
+    elif getattr(rep2, '_differentials', False):
+        return False
+
+    return result & components_equal(rep1, rep2)
```
These are copy-pasted from `test_representations.py`. Should we move them to `astropy.utils` or `astropy.tests.helper`, or is it not too bad to have them duplicated?

mhvk replied:
I'd just import then from `test_representation` for now (also done in `test_representation_methods`...)

astrojuanlu replied:
Done!

astrojuanlu commented:
Addressed the first batch of comments, going to look into `EarthLocation` now.

astrojuanlu commented:
What I did in the end was to deprecate `.to_geodetic` and `.to_geocentric` without changing their return values, and added a new `.represent_as`.

This is ready for a second pass!

astrojuanlu commented:
Many test failures due to the `AstropyDeprecationWarning` I introduced - I guess no internal calls to deprecated methods are allowed. Will fix it.

mkbrewer commented:
I would just note that many locations on Earth have heights that are below that of the ellipsoid. I seems a bit unreasonable to force a different method for creating and `EarthLocation` in these instances.

mhvk commented:
Darn, we'd need `allow_negative`. I think for now it would be best to just proceed with a `Quantity` - we can deal with insisting it is a length differently/later.

mhvk commented on astropy/coordinates/earth.py:
```diff
@@ -595,14 +593,28 @@ def to_geodetic(self, ellipsoid=None):
         For the conversion to geodetic coordinates, the ERFA routine
         ``gc2gd`` is used.  See https://github.com/liberfa/erfa
         """
+        warn("Method .to_geodetic(ellipsoid) is deprecated, "
```
Agree with the deprecation message, but in analogy with a frame or sky coord, I think there is little against still having the properties for particular representations, i.e., I think we should allow (and also refer to in the warning message),
```
el.{wgs84,...}geodetic
el.cartesian
```

astrojuanlu replied:
Removed the deprecation message for now, but saving this for the future.

mhvk commented on astropy/coordinates/earth.py:
```diff
@@ -595,14 +593,28 @@ def to_geodetic(self, ellipsoid=None):
+            llh.lat,
+            u.Quantity(llh.height, self.unit, copy=False))
+
+    def represent_as(self, other_class):
```
Again seeing `EarthLocation` as more like a frame/skycoord rather than a representation, I think we can allow strings. I'd suggest following the nomenclature in their `represent_as` method and calling it `base`.

astrojuanlu replied:
Removed `represent_as` for now, but saving this for the future.

mhvk commented on astropy/coordinates/earth.py:
```diff
@@ -848,3 +864,84 @@ def _to_value(self, unit, equivalencies=[]):
+
+    Parameters
+    ----------
+    lon, lat : `~astropy.units.Quantity`
```
The classes are `Longitude` and `Latitude` - might as well be precise...

astrojuanlu replied:
Tried this trick https://github.com/astropy/astropy/issues/11063#issuecomment-732212333 let's see if it works!

mhvk commented on astropy/coordinates/earth.py:
```diff
@@ -848,3 +864,84 @@ def _to_value(self, unit, equivalencies=[]):
         return new_array.view(self.dtype).reshape(self.shape)
+
+
+class BaseGeodeticRepresentation(BaseRepresentation):
```
I think this will get registered in `REPRESENTATION_CLASSES`, which is not ideal... Pinging @adrn for ideas; should we allow a `name = False` explicit way to avoid registration? (Note: this should not have to hold up this PR!)

mhvk commented on astropy/coordinates/earth.py:
```diff
@@ -848,3 +864,84 @@ def _to_value(self, unit, equivalencies=[]):
+
+
+class WGS84GeodeticRepresentation(BaseGeodeticRepresentation):
+    """
```
One disadvantage of setting this docstring is that you lose the nice information you added in the `Base...`.  For a possible solution, see https://docs.astropy.org/en/latest/api/astropy.utils.decorators.format_doc.html#astropy.utils.decorators.format_doc

astrojuanlu replied:
I ended up removing the docstrings, after all the name should be explicit enough.

However, I did a local `tox -e build_docs` and I didn't see these classes anywhere in the docs, so I added them to `__all__` in `earth.py`.

astrojuanlu replied:
I was wrong, removing the docstrings leaves the class documentation empty:

![Screenshot_2020-12-01 WGS84GeodeticRepresentation — Astropy v4 3 dev322+g7f34535cc d20201201](https://user-images.githubusercontent.com/316517/100720080-c57bea00-33bd-11eb-83a8-ca18579e170e.png)

astrojuanlu replied:
`@format_doc` proved to be the best solution :+1:

mhvk commented on astropy/coordinates/tests/test_earth.py:
```diff
@@ -410,3 +437,236 @@ def test_info():
     assert str(greenwich.info).startswith('name = Royal Observatory Greenwich')
+
+
+class TestWGS84GeodeticRepresentationRepresentation:
```
I would shrink this class substantially - many of the tests are done on `Spherical` to just have them done once, no need to do them again. Also, do re-align for the longer name...

mhvk replied:
Thinking more, since all new tests are (very nicely) just focussed on the representations, how about a new `test_geodetic_representations`?

astrojuanlu replied:
I moved these tests to `test_geodetic_representations.py`, but I'm not sure which ones to remove.

mhvk commented on astropy/coordinates/tests/test_earth.py:
```diff
@@ -410,3 +437,236 @@ def test_info():
+        assert quantity_allclose(new_sph.lon, sph.lon)
+        assert quantity_allclose(new_sph.lat, sph.lat)
+
+        # dif = SphericalCosLatDifferential(1*u.mas/u.yr, 2*u.mas/u.yr,
```
Probably good to put a TODO on top that we do not yet have differentials (and then remove the commented out code here)

astrojuanlu replied:
Done!

mhvk commented on astropy/coordinates/tests/test_earth.py:
```diff
@@ -410,3 +437,236 @@ def test_info():
+
+    assert quantity_allclose(gc.x, -5599000.5577049947 * u.m, atol=1e-7 * u.m)
+    assert quantity_allclose(gc.y, 233011.67223479203 * u.m, atol=1e-7 * u.m)
+    assert quantity_allclose(gc.z, -3040909.4706983363 * u.m, atol=1e-7 * u.m)
```
Missing from the tests is the new `represent_as` method.

astrojuanlu replied:
Removed `represent_as` for now, will test it when we finally add it.

mhvk reviewed (commented):
@astrojuanlu - this is starting to look very nice.  Bit of a mix of larger and probably overly detailed comments.

More general one is where we stop. One option is actually a bit *before* where you are now, i.e., not add deprecation messages yet, but just introduce the new representations and use them *internally* only for the transformations. This has the advantage that no tests will fail yet. Arguably, in that case, we would also not yet add the `represent_as` method (which is not tested yet, so that is fine, I think).

The other is perhaps a bit further than you are now, whether `EarthLocation` stops becoming a `Quantity` and instead stores the representation in `.data` like a frame (or perhaps as an intermediate suggestion still shares the data with a `CartesianRepresentation`).

But my sense would be to take the smaller step suggested above, so we can get this in and follow-up on it.

astrojuanlu commented:
> But my sense would be to take the smaller step suggested above, so we can get this in and follow-up on it.

I agree - I will remove the deprecation warnings and remove `represent_as` for now, and follow up in a future PR. Addressing your comments now.

mkbrewer commented:
While supporting alternate ellipsoids is fine, you should realize that there is more work involved in doing so. As currently written, `ErfaAstrom.apco()` calls `erfa.apco()` which calculates the observer's position and velocity based on the WGS-84 ellipsoid. Thus geodetic  coordinates based on an alternate ellipsoid will yield incorrect results in any ICRS<->CIRS or ICRS<->AltAz transform.

mkbrewer commented:
I should note that this relates to task 1) in issue #11072

astrojuanlu commented:
Warnings are legitimate:

```
/home/docs/checkouts/readthedocs.org/user_builds/astropy/envs/11086/lib/python3.8/site-packages/astropy/coordinates/earth.py:docstring of astropy.coordinates.GRS80GeodeticRepresentation:1: WARNING: py:class reference target not found: astropy.coordinates.earth.BaseGeodeticRepresentation
writing output... [ 16%] api/astropy.coordinates.SphericalCosLatDifferential .. api/astropy.io.ascii.Latex
/home/docs/checkouts/readthedocs.org/user_builds/astropy/envs/11086/lib/python3.8/site-packages/astropy/coordinates/earth.py:docstring of astropy.coordinates.WGS72GeodeticRepresentation:1: WARNING: py:class reference target not found: astropy.coordinates.earth.BaseGeodeticRepresentation
/home/docs/checkouts/readthedocs.org/user_builds/astropy/envs/11086/lib/python3.8/site-packages/astropy/coordinates/earth.py:docstring of astropy.coordinates.WGS84GeodeticRepresentation:1: WARNING: py:class reference target not found: astropy.coordinates.earth.BaseGeodeticRepresentation
...
[changelog_links] Adding changelog links to "changelog"
<rst_epilog>:14: WARNING: py:class reference target not found: astropy.coordinates.earth.BaseGeodeticRepresentation
```

fixing

astrojuanlu commented:
> As currently written, `ErfaAstrom.apco()` calls `erfa.apco()` which calculates the observer's position and velocity based on the WGS-84 ellipsoid. Thus geodetic coordinates based on an alternate ellipsoid will yield incorrect results in any ICRS<->CIRS or ICRS<->AltAz transform.

I don't understand @mkbrewer , isn't `ErfaAstrom.apco()` always converting whatever coordinates to WGS84?

https://github.com/astropy/astropy/blob/9680cd546aa9063758f2c23c836ca79a7c8f1eb1/astropy/coordinates/erfa_astrom.py#L45

Unless I missed anything, the `EarthLocation.to_geodetic` return value didn't change with this PR.

astrojuanlu commented:
All CI checks passed now :tada:

mkbrewer commented:
> I don't understand @mkbrewer , isn't ErfaAstrom.apco() always converting whatever coordinates to WGS84?

Oh, you're right. I missed that conversion. Sorry.

mhvk commented on CHANGES.rst:
```diff
@@ -23,6 +23,9 @@ astropy.coordinates
   provides a modest speedup of approximately 10 percent. [#11079]
 
+- Adds new ``WGS84GeodeticRepresentation``, ``WGS72GeodeticRepresentation``,
+  and ``GRS80GeodeticRepresentation``. [#11086]
```
Should we note what these are good for? E.g., "These are mostly for use inside ``EarthLocation`` but can also be used to convert between geocentric (cartesian) and different geodetic representations directly."

mhvk commented on astropy/coordinates/earth.py:
```diff
@@ -287,22 +291,18 @@ def from_geodetic(cls, lon, lat, height=0., ellipsoid=None):
+        xyz = REPRESENTATION_CLASSES[ellipsoid.lower() + "geodetic"](
+            lon, lat, Distance(height, allow_negative=True), copy=False,
+        ).represent_as(CartesianRepresentation).get_xyz(xyz_axis=-1)
         self = xyz.ravel().view(cls._location_dtype,
```
Might be clearer to change the unit of `xyz` here instead of `self` at the end - mostly as we can do it in-place.

mhvk commented on astropy/coordinates/earth.py:
```diff
@@ -596,13 +596,12 @@ def to_geodetic(self, ellipsoid=None):
-            u.Quantity(height * u.meter, self.unit, copy=False))
+            llh.lon,
+            llh.lat,
+            u.Quantity(llh.height, self.unit, copy=False))
```
Can just be `llh.height << self.unit`.

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -1810,6 +1810,8 @@ def __init__(self, lon, lat=None, distance=None, differentials=None,
             try:
                 self._distance = Distance(self._distance, copy=False)
             except ValueError as e:
+                # FIXME: This will raise no matter what,
```
Maybe keep outside of this PR?

mhvk reviewed (commented):
This is looking good! It made me realize that really `gc2gd` and `gd2gc` should be made to work with `Quantity` - so I took the liberty of adding a commit that does so.

Also, many of the other comments are a bit nit-picky, so while I was at it, I addressed them. Main thing to perhaps think about is that I removed the use of `Distance` - I find it just too unhandy, not really suitable for a height, and it really is more useful for converting to redshift, etc...

Will push those 2 commits to your branch...

mhvk commented:
Oops, some failures - will fix (should have tested all of `coordinates`...

astrojuanlu commented:
Thanks @mhvk ! I can leave the rest to you or address the comments tomorrow, as you prefer

mhvk commented on astropy/coordinates/earth.py:
```diff
@@ -848,3 +847,82 @@ def _to_value(self, unit, equivalencies=[]):
+class BaseGeodeticRepresentation(BaseRepresentation):
+    """Base geodetic representation."""
+
+    attr_classes = {'lon': Longitude,
```
Let's leave this for follow-up...

mhvk commented:
@astrojuanlu - OK, I also changed the tests now. Plus one thing that you should sign off on: to make `ELLIPSOIDS` a dict with the actual representations as values. I think it works quite nicely, mostly because #11090 already replaced the meta class with `__init_subclass__`, so one can now easily chain things. See 4612ca1.

mhvk reviewed (approved):
Approving, but, as noted, @astrojuanlu, would be best if you approve the changes I made!

astrojuanlu commented on astropy/coordinates/earth.py:
```diff
@@ -875,6 +876,11 @@ class BaseGeodeticRepresentation(BaseRepresentation):
                     'lat': Latitude,
                     'height': u.Quantity}
 
+    def __init_subclass__(cls, **kwargs):
```
Nice trick!

astrojuanlu reviewed (commented):
Can't approve my own pull request. This looks good to me, thanks @mhvk for the final tweaks!
